src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,731 @@
+            return Task.FromException<HttpResponseMessage>(new HttpRequestException("Server sent GOAWAY.", null, RequestRetryType.RetryOnSameOrNextProxy));
+        }
+
+        /// <summary>
+        /// Waits for MAX_STREAMS to be raised by the server.
+        /// </summary>
+        private Task WaitForAvailableRequestStreamAsync(CancellationToken cancellationToken)
+        {
+            TaskCompletionSourceWithCancellation<bool> tcs;
+
+            lock (SyncObj)
+            {
+                long remaining = _requestStreamsRemaining;
+
+                if (remaining > 0)
+                {
+                    _requestStreamsRemaining = remaining - 1;
+                    return Task.CompletedTask;
+                }
+
+                tcs = new TaskCompletionSourceWithCancellation<bool>();
+                _waitingRequests.Enqueue(tcs);
+            }
+
+            // Note: cancellation on connection shutdown is handled in CancelWaiters.
+            return tcs.WaitWithCancellationAsync(cancellationToken).AsTask();
+        }
+
+        /// <summary>
+        /// Cancels any waiting SendAsync calls.
+        /// </summary>
+        /// <remarks>Requires <see cref="SyncObj"/> to be held.</remarks>
+        private void CancelWaiters()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObj));
+
+            while (_waitingRequests.TryDequeue(out TaskCompletionSourceWithCancellation<bool> tcs))
+            {
+                tcs.TrySetException(new HttpRequestException("Server sent GOAWAY or aborted.", null, RequestRetryType.RetryOnSameOrNextProxy));
+            }
+        }
+
+        // TODO: how do we get this event?
+        private void OnMaximumStreamCountIncrease(long newMaximumStreamCount)
+        {
+            lock (SyncObj)
+            {
+                if (newMaximumStreamCount <= _maximumRequestStreams)
+                {
+                    return;
+                }
+
+                _requestStreamsRemaining += (newMaximumStreamCount - _maximumRequestStreams);
+                _maximumRequestStreams = newMaximumStreamCount;
+
+                while (_requestStreamsRemaining != 0 && _waitingRequests.TryDequeue(out TaskCompletionSourceWithCancellation<bool> tcs))
+                {
+                    if (tcs.TrySetResult(true))
+                    {
+                        --_requestStreamsRemaining;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Aborts the connection with an error.
+        /// </summary>
+        /// <remarks>
+        /// Used for e.g. I/O or connection-level frame parsing errors.
+        /// </remarks>
+        internal void Abort(Exception abortException)
+        {
+            // Only observe the first exception we get.
+            Exception firstException = Interlocked.CompareExchange(ref _abortException, abortException, null);
+
+            if (firstException != null &&
+                NetEventSource.IsEnabled &&
```

Comment:
I think I'm confused on the purpose of NetEventSource. Is this something customers could/would have enabled which they could see? I think it's fine to log something like this, but I'm also not a fan of intertwining NetEventSource checks and actual logic for the method.

Also, should we always return even if NetEventSource isn't enabled?

Reply:
> I think I'm confused on the purpose of NetEventSource. Is this something customers could/would have enabled which they could see?

Yes, it is a log you can get at from event viewer.

> Also, should we always return even if NetEventSource isn't enabled?

Good catch.